File: Modules/WillStrohl.Injection/Components/DataProvider.vb
'
' Will Strohl (will.strohl@gmail.com)
' http://www.willstrohl.com
'
'Copyright (c) 2009-2015, Will Strohl
'All rights reserved.
'
'Redistribution and use in source and binary forms, with or without modification, are 
'permitted provided that the following conditions are met:
'
'Redistributions of source code must retain the above copyright notice, this list of 
'conditions and the following disclaimer.
'
'Redistributions in binary form must reproduce the above copyright notice, this list 
'of conditions and the following disclaimer in the documentation and/or other 
'materials provided with the distribution.
'
'Neither the name of Will Strohl, Content Injection, nor the names of its contributors may be 
'used to endorse or promote products derived from this software without specific prior 
'written permission.
'
'THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY 
'EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES 
'OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT 
'SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
'INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED 
'TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR 
'BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
'CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
'ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH 
'DAMAGE.
'

Imports DotNetNuke

Namespace WillStrohl.Modules.Injection

    Public MustInherit Class DataProvider

#Region " Private Members "

        Private Const c_AssemblyName As String = "WillStrohl.Modules.Injection.SqlDataProvider, WillStrohl.Modules.Injection"

#End Region

#Region " Shared/Static Methods "

        ' singleton reference to the instantiated object 
        Private Shared objProvider As DataProvider = Nothing

        ' constructor
        Shared Sub New()
            CreateProvider()
        End Sub

        ' dynamically create provider
        Private Shared Sub CreateProvider()
            'objProvider = CType(Framework.Reflection.CreateObject(c_Object, c_ObjectType, c_AssemblyName), DataProvider)

            If objProvider Is Nothing Then
                Dim objectType As Type = Type.GetType(c_AssemblyName)

                objProvider = DirectCast(Activator.CreateInstance(objectType), DataProvider)
                DataCache.SetCache(objectType.FullName, objProvider)
            End If
        End Sub

        ' return the provider
        Public Shared Shadows Function Instance() As DataProvider
            If objProvider Is Nothing Then
                CreateProvider()
            End If
            Return objProvider
        End Function

#End Region

#Region " Abstract Methods "

        Public MustOverride Function AddInjectionContent(ByVal ModuleId As Integer, ByVal InjectTop As Boolean, ByVal InjectName As String, ByVal InjectContent As String, ByVal IsEnabled As Boolean, ByVal OrderShown As Integer) As Integer
        Public MustOverride Sub UpdateInjectionContent(ByVal InjectionId As Integer, ByVal ModuleId As Integer, ByVal InjectTop As Boolean, ByVal InjectName As String, ByVal InjectContent As String, ByVal IsEnabled As Boolean, ByVal OrderShown As Integer)
        Public MustOverride Sub DisableInjectionContent(ByVal InjectionId As Integer)
        Public MustOverride Sub EnableInjectionContent(ByVal InjectionId As Integer)
        Public MustOverride Sub DeleteInjectionContent(ByVal InjectionId As Integer)
        Public MustOverride Function GetInjectionContent(ByVal InjectionId As Integer) As IDataReader
        Public MustOverride Function GetActiveInjectionContents(ByVal ModuleId As Integer) As IDataReader
        Public MustOverride Function GetInjectionContents(ByVal ModuleId As Integer) As IDataReader
        Public MustOverride Function GetNextOrderNumber(ByVal ModuleId As Integer) As Integer
        Public MustOverride Sub ChangeOrder(ByVal InjectionId As Integer, ByVal Direction As String)
        Public MustOverride Function DoesInjectionNameExist(ByVal InjectionName As String, ByVal ModuleId As Integer) As Boolean

#End Region

    End Class

End Namespace

File: Modules/WillStrohl.Injection/Components/DataProvider.cs
/*
Copyright © Upendo Ventures, LLC

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and 
associated documentation files (the "Software"), to deal in the Software without restriction, 
including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, 
and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial 
portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT 
NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES 
OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN 
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

using DotNetNuke.Common.Utilities;
using System;
using System.Data;

namespace WillStrohl.Modules.Injection.Components
{
	public abstract class DataProvider
	{
		#region " Private Members "
        
		private const string c_AssemblyName = "WillStrohl.Modules.Injection.Components.SqlDataProvider, WillStrohl.Modules.Injection";

		#endregion

		#region " Shared/Static Methods "

		// singleton reference to the instantiated object 
        private static DataProvider objProvider = null;
		
        // constructor
		static DataProvider()
		{
			CreateProvider();
		}

		// dynamically create provider
		private static void CreateProvider()
		{
			if (objProvider != null) return;
		    
            var objectType = Type.GetType(c_AssemblyName);

		    objProvider = (DataProvider)Activator.CreateInstance(objectType);

		    DataCache.SetCache(objectType.FullName, objProvider);
		}

		// return the provider
		public static new DataProvider Instance()
		{
			if (objProvider == null) {
				CreateProvider();
			}
			return objProvider;
		}

		#endregion

		#region " Abstract Methods "

		public abstract int AddInjectionContent(int ModuleId, bool InjectTop, string InjectName, string InjectContent, bool IsEnabled, int OrderShown, string CustomProperties);
		public abstract void UpdateInjectionContent(int InjectionId, int ModuleId, bool InjectTop, string InjectName, string InjectContent, bool IsEnabled, int OrderShown, string CustomProperties);
		public abstract void DisableInjectionContent(int InjectionId);
		public abstract void EnableInjectionContent(int InjectionId);
		public abstract void DeleteInjectionContent(int InjectionId);
		public abstract IDataReader GetInjectionContent(int InjectionId);
		public abstract IDataReader GetActiveInjectionContents(int ModuleId);
		public abstract IDataReader GetInjectionContents(int ModuleId);
		public abstract int GetNextOrderNumber(int ModuleId);
		public abstract void ChangeOrder(int InjectionId, string Direction);
		public abstract bool DoesInjectionNameExist(string InjectionName, int ModuleId);

		#endregion
	}
}